Text proposed to address scope/expectation (#541)
quicwg/multipath: draft-ietf-quic-multipath.md

to be handled by the application. But this document does not prevent future extensions from
defining mechanisms to cope with the remaining scenarios.

> This extension can be used with 
> different scheduling algorithms that define
> how multiple simultaneously open paths are used to send packets. 
> Examples of schedulers can range from support for failover to simulatenous
> use of the aggreate capacity across all active paths.
> 
> This document does not specify a scheduling algorithm.
> There are IETF specifications for path fail-over, but
> there are currently no IETF standards-track specifications for simultaneously
> (concurrently) using the aggreate capacity of a set of active paths. 
> 
> Because the scheduling policy differs depending on application requirements,
> only some basic implementation guidance is discussed in {{impl-consideration}}.

Reviewer: Thanks for further updates.

Proposing again to re-order your sentences to keep changes to the existing text minimal:
```suggestion
Further, this document does not specify scheduling algorithms that define
how multiple, simultaneously open paths are used to send packets.
As these differ depending on application requirements,
only some basic implementation guidance is discussed in {{impl-consideration}}.
This extension can be used with different scheduling algorithms that,
e.g., can range from support for failover to simulatenous
use of the aggreate capacity across all active paths.
There are IETF specifications for path fail-over, but
there are currently no IETF standards-track specifications for simultaneously
(concurrently) using the aggregated capacity of a set of active paths. 
```

to be handled by the application. But this document does not prevent future extensions from
defining mechanisms to cope with the remaining scenarios.

> This extension can be used with 
> different scheduling algorithms that define
> how multiple simultaneously open paths are used to send packets. 
> Examples of schedulers can range from support for failover to simulatenous
> use of the aggreate capacity across all active paths.
> 
> This document does not specify a scheduling algorithm for multipath in
> the general Internet.
> There are IETF specifications for path fail-over, but
> there are currently no IETF standards-track specifications for simultaneously
> (concurrently) using the aggreate capacity of a set of active paths. 
> 
> Because the scheduling policy differs depending on application requirements,
> only some basic implementation guidance is discussed in {{impl-consideration}}.

Reviewer: This is a proposal to rearrange the text to make the different to the existing text smaller:
```suggestion
Further, this document does not specify scheduling algorithms 
for multipath usage in the general Internet.
As these differ depending on application requirements,
only some basic implementation guidance is discussed in {{impl-consideration}}.
This extension can be used with  different scheduling algorithms that
e.g. can range from support for failover to simultaneous
use of the aggregated capacity across all active paths.
There are IETF specifications for path fail-over, but
there are currently no IETF standards-track specifications for simultaneously
(concurrently) using the aggreate capacity of a set of active paths. 
```
However, I still have some comments on this text:
a) I don't think we should say "scheduling for the general Internet" because this document does actually not specify any scheduling algorithm. If you add "for the general Internet" this sounds like it does specify something but it can only not be used in the general Internet. Why is that needed?
b) I would like to still remove "standards track" because I don't think it matter if that doc would be PS or informational and also this sound like there is only no PR doc but there is no doc at all.
c) If we say there are specs for failover we should add a reference, however, as I said earlier I don't think RFC7829 is really relevant here.
d) I would still prefer a shorter version as proposed by my previously

Author: (a) Changed. 
(b, c) The principle behind RFC7829 might or might not be relevant - that would be for the WG to decide. If the WG thinks it is not relevant background for QUIC, then the I-D could say there are no current IETF standards-track documents defining schedulers for MP_QUIC for use in the general Internet.
- I'd much prefer text that also includes something like "if and when the IETF standardises a scheduling method this can be used with this extension".
(d) Of course, you might prefer your own shorter text;-). 

I will anyway be sure to look to what has been finally written, and will ask questions when the WG decides to request publication.

to be handled by the application. But this document does not prevent future extensions from
defining mechanisms to cope with the remaining scenarios.

The framework specified in this document can be used with 
a range of scheduling algorithms that define
how multiple simultaneously open paths are used to send packets. 
Examples of schedulers can range from support for failover to simulatenous concurrent
use of the aggreate capacity across all active paths.

The specification of scheduling for multipath in
the general Internet is outside the scope of this document.  
There are existing IETF specifications for path fail-over, but

> concurrently no IETF standards-track specifications for simulatenously

Reviewer: ```suggestion
concurrently no IETF standards-track specifications for simultaneously
```

using the capacity of a set of active paths. 

> Because the scheduling policy differs depending on application requirements,

Reviewer: A couple fo comments:

- This text again used the term "framework" which we don't use in this document and I don't find appropriate.
- I personally prefer the phrasing "this document does not specify" instead of talking about "out of scope". I think that is more clear and otherwise the question comes up why is it out of scope. I think it's reflecting more how we decided to focus the document and leave other things for future work. Why do you think that the "out-of-scope" re-wording is needed/different/better?
- Are there existing IETF specification of path fail-over? Which? if so we should add a reference somewhere (not necessarily here but maybe in the scheduling section) but not sure what you mean.
- I don't think we need to say "simultaneously using the capacity of a set of" that is just more words and still say the same thing as "simultaneously using multiple paths".
- Actually I don't really understand why it is important to mention "standards-track" here?
- Saying "simultaneous concurrent" is saying the same thing twice. I don't see a different here. We keep using "simultaneous" in this document, so I think we should stick with it. We could change that but I don't see a difference.

> only some basic implementation guidance is discussed in {{impl-consideration}}.

Reviewer: A couple fo comments:

- This text again used the term "framework" which we don't use in this document and I don't find appropriate.
- I personally prefer the phrasing "this document does not specify" instead of talking about "out of scope". I think that is more clear and otherwise the question comes up why is it out of scope. I think it's reflecting more how we decided to focus the document and leave other things for future work. Why do you think that the "out-of-scope" re-wording is needed/different/better?
- Are there existing IETF specification of path fail-over? Which? if so we should add a reference somewhere (not necessarily here but maybe in the scheduling section) but not sure what you mean.
- I don't think we need to say "simultaneously using the capacity of a set of" that is just more words and still say the same thing as "simultaneously using multiple paths".
- Actually I don't really understand why it is important to mention "standards-track" here?
- Saying "simultaneous concurrent" is saying the same thing twice. I don't see a different here. We keep using "simultaneous" in this document, so I think we should stick with it. We could change that but I don't see a difference.

Reviewer: How about keep the old text and adding the following sentence, which I believes covers what you want to say above with less wordings:

```suggestion
Further, this document does not specify scheduling algorithms that define
how multiple, simultaneously open paths are used to send packets.
As these differ depending on application requirements,
only some basic implementation guidance is discussed in {{impl-consideration}}.
However, this extension can be used with different scheduling algorithms,
e.g., for failover or capacity aggregation, even though
there is no IETF document that specifies scheduling algorithms
for simultaneous path usage at this point of time.
```

Author: I made a new suggestion, based on the above, see PR text.

Reviewer: Thanks for the reference to RFC7829. I wasn't aware of this RFC. However, I’m not sure RFC7829 is really relevant here because I don’t think it necessarily assumes that packet are transmitted on more than one path at any point of time. It’s more you want to detect a breakage quickly and stop sending on one path and then start sending on another path. So in other words no concurrent use. However, I only had a quick look.

On the term concurrent multipath: MP-TCP uses “simultaneous". There is not a single occurrence of “concurrent” in that RFC8684. So why is this a requirement for this document now? Also we now refer to RFC9743 which explains and uses that term but given scheduling is not part of this document, I don’t see why we need to use it. I think “simultaneous” is equally clear. 

However, here is a proposal to add this to my proposed sentence above:
```suggestion
Further, this document does not specify scheduling algorithms that define
how multiple, simultaneously open paths are used to send packets.
As these differ depending on application requirements,
only some basic implementation guidance is discussed in {{impl-consideration}}.
This extension can be used with different scheduling algorithms,
e.g., for failover or capacity aggregation, even though
there is no IETF document that specifies scheduling algorithms
for simultaneous path usage, sometime also referred to as concurrent multipath, at this point of time.
```

@Gorry please let me know if that proposed text/sentence would be okay for you instead of your text?

However, I would also like to get more input from others before merging! @LPardue @huitema @qdeconinck @Yanmei-Liu

Specifically, while failover between Wi-Fi
and mobile networks is a well-known multipath use case,
it only temporarily uses two paths at the same time 
to avoid transmission pauses.

> More general use of simultaneous paths, however, needs more consideration

Author: I think we differ on what "generally" implies. This isn't important to me.

> More general use of simultaneous paths, however, needs more consideration

Author: ... You can write: Simultaneous path usage generally, however, needs more consideration.

> More general use of simultaneous paths, however, needs more consideration

Reviewer: As agreed reverting to original text:
```suggestion
Simultaneous path usage generally, however, needs more consideration
```

than specified in this document to avoid negative performance
impacts, e.g., when stream data is distributed over multiple paths with

> different delays or to avoid excessive conestion of shared bottlenecks.

Reviewer: ```suggestion
different delays or to avoid excessive congestion of shared bottlenecks.
```

> different delays or to avoid excessive congestion of shared bottlenecks.

Reviewer: This is a separate issue and not necessarily a "negative performance impact". E.g. early version of BBR could lead to high loss but exactly because of its aggressiveness would provide better application performance.
Also congestion control is there to avoid excessive congestion and this has nothing to do with use of multipath. So what is the actually problem for shared bottlenecks that you want to point out here?

Author: It's an example, and if you think there are sensitivities you can skip the second example.
The example is about avoiding *excessive congestion*, and that can arise, or can not arise when  you share a common bottleneck - I'd say congestion is not a negative impact but excessive congestion is. 
Anyway, as I noted I REALLY don't mind not having the example.

Reviewer: sorry I typed my second reply on an out-dated view where Gorry's reply didn't show up and as such our postings crossed each other. I'm okay with just remove it. Thanks!

Reviewer: As agreed removing this additional text:
```suggestion
different delays.
```
